src/web/core/src/main/java/org/geoserver/web/security/AccessDataRuleInfoManager.java

```diff
@@ -0,0 +1,220 @@
+                                    }
+                                });
+            }
+            modeRoleMap.put(auth, modes);
+        }
+        rules.forEach(r -> authorities.removeAll(r.getRoles()));
+        authorities.forEach(r -> modeRoleMap.put(r, null));
+        for (String key : modeRoleMap.keySet()) {
+            Set<AccessMode> ams = modeRoleMap.get(key);
+            DataAccessRuleInfo model = new DataAccessRuleInfo(key, wsName, layerName);
+            model.setAdminFromMode(ams);
+            model.setReadFromMode(ams);
+            model.setWriteFromMode(ams);
+            models.add(model);
+        }
+        return models;
+    }
+
+    public List<DataAccessRuleInfo> getNewInfoList(
+            String wsName, String layerName, Set<String> availableRoles) {
+        List<DataAccessRuleInfo> rules = new ArrayList<>();
+        for (String role : availableRoles) {
+            DataAccessRuleInfo rule = new DataAccessRuleInfo(role, wsName, layerName);
+            rule.setRead(false);
+            rule.setWrite(false);
+            rule.setAdmin(false);
+            rules.add(rule);
+        }
+        return rules;
+    }
+
+    public Set<DataAccessRule> mapFrom(
+            List<DataAccessRuleInfo> newRules,
+            Set<String> authorities,
+            String wsName,
+            String layerName,
+            boolean globalLayerGroup) {
+
+        Set<DataAccessRule> rules = new HashSet<>(authorities.size());
+        Map<AccessMode, Set<String>> modeRoleMap = new HashMap<>(MODES.size());
+        for (AccessMode mode : MODES) {
+            Set<String> selectedRoles = new HashSet<>();
+            for (String auth : authorities) {
+                newRules.stream()
+                        .filter(role -> role.getRoleName().equalsIgnoreCase(auth))
+                        .forEach(
+                                rule -> {
+                                    if (rule.hasMode(mode)) selectedRoles.add(auth);
+                                });
+            }
+            modeRoleMap.put(mode, selectedRoles);
+        }
+        for (AccessMode key : modeRoleMap.keySet()) {
+            Set<String> roles = modeRoleMap.get(key);
+            if (roles != null && roles.size() > 0) {
+                DataAccessRule rule = new DataAccessRule();
+                if (!globalLayerGroup) {
+                    rule.setRoot(wsName);
+                    rule.setLayer(layerName != null ? layerName : "*");
+                } else {
+                    rule.setRoot(layerName);
+                    rule.setLayer(null);
+                    rule.setGlobalGroupRule(true);
+                }
+                rule.setAccessMode(key);
+                rule.getRoles().addAll(roles);
+                rules.add(rule);
+            }
+        }
+        return rules;
+    }
+
+    public void saveRules(Set<DataAccessRule> old, Set<DataAccessRule> news) throws IOException {
```

Comment:
Thinking a bit more, the GUI is under the global lock to avoid concurrent modifications to the catalog... but it is possible to disable it... some level of synch is probably needed anyways.